(SERVER-303) Remove version-check service

fpringvaldsen opened the pull request:
Remove the dujour version-check service and replace it with the
dujour-version-check library.

I've flagged this as for review as I needed to add a new function, `get-version-string`, to the `dujour-version-check` library for this to work. That change is present in https://github.com/puppetlabs/dujour-version-check/pull/7. Once that is merged, we can do a new release of the `dujour-version-check` library and update this from a snapshot dependency.

camlow325 commented on src/clj/puppetlabs/services/master/master_service.clj:
```diff
@@ -4,25 +4,31 @@
+         product-name      (get-in-config [:product :name])
+         update-server-url (get-in-config [:product :update-server-url])]
+
+     (version-check/check-for-updates! {:product-name product-name} update-server-url)
```
Do we have a test which validates that the master service makes this call with the right parameters?  If not, it would seem good to create one.  Maybe could stub out the call to `check-for-updates!` with a `with-redefs` so that the full plumbing doesn't have to be exercised for this local unit test?

fpringvaldsen replied:
We do not, all we had before were tests of the various functions in the `version-check-core` namespace, which is now covered by the tests in the `dujour-version-check` library.

I'll work on adding something to ensure that the `check-for-updates!` function is successfully being called with the right parameters.

fpringvaldsen replied:
@camlow325 I've added a test as requested.

camlow325 commented:
:+1: with a question on an additional unit test, assuming the dujour-version-check dependency is moved off of a SNAPSHOT, and Travis tests subsequently pass.

fpringvaldsen commented on test/unit/puppetlabs/services/master/master_service_test.clj:
```diff
@@ -70,3 +69,56 @@
+    (with-redefs
+
+      [version-check/check-for-updates! version-check-test-fn
+       ; Redefine the CA functions to ensure that the master does not create
```
I did this since it stops the master service from creating SSL files, which sped up the test a bit. I also figured it would reduce the amount of things the master is doing on initialization since all we care about in this test is whether or not it calls into the `check-for-updates!` function. Further, this eliminates the need to manage a test directory in the test.

I'm happy to remove this though, and manage a test directory in the above test.

camlow325 replied:
I'm good with mocking these here since this functionality isn't related to the test.

camlow325 commented on test/unit/puppetlabs/services/master/master_service_test.clj:
```diff
@@ -70,3 +69,50 @@
+  [request-values update-server-url]
+  (is (= {:group-id "puppets" :artifact-id "yoda"} (:product-name request-values)))
+  (is (= "http://notarealurl/" update-server-url))
+  (swap! version-check-success not))
```
This is purely a style comment, so feel free to ignore.  I'd be inclined to have this just set the atom with the contents of the arguments passed into the function - maybe as a map with a key/val pair for each argument - and move these `is` assertions into the body of the deftest.  Where possible, I like to have assertions inside the body of the `deftest` so that what it is testing is more prominent.

camlow325 replied:
This test otherwise looks good to me.

fpringvaldsen commented:
@camlow325 I've moved the assertions into the test itself, and am now using the atom to store the values of the parameters passed to `version-check-test-fn`.

cprice404 commented:
Philosophical ranting aside... +1

fpringvaldsen commented:
@cprice404 @camlow325 I've moved the atom and function into a `let` block in the version check test. I've also bumped the dependency on `dujour-version-check` so this should be ready to go now.

camlow325 commented:
:+1: Looks like Travis failed.  Not sure if that's because it was down for maintenance?

cprice404 commented:
kicked travis

fpringvaldsen commented:
@cprice404 @camlow325 I released `dujour-version-check` to clojars and moved the dependency out of the `dev` profile and into the main dependencies list. Hopefully travis will pass now.

fpringvaldsen commented:
Okay I think I've finally fixed the Travis failures, we'll see how this new commit does.

fpringvaldsen commented:
@cprice404 @camlow325 It looks like Travis is finally fixed. I've made one new commit, https://github.com/puppetlabs/puppet-server/commit/fe08e87856a748c2a4d7738eb7f236cd5a91a0b6, to fix an issue with the `version-check-test`.

cprice404 commented:
@jeffmccune I presume that you wouldn't be aversely affected by us going ahead and merging this?

nwolfe commented on src/clj/puppetlabs/services/master/master_service.clj:
```diff
@@ -4,25 +4,31 @@
    [:CaService initialize-master-ssl! retrieve-ca-cert! retrieve-ca-crl!]
-   [:JRubyPuppetService]]
+   [:JRubyPuppetService]
+   [:ConfigService get-in-config]]
```
This already depends on the `PuppetServerConfigService`, can we just get the config from that?  The entire `config` is pulled in here: https://github.com/puppetlabs/puppet-server/pull/409/files#diff-f69c98ef06ea1897ed3b68e38fb638d3R21

(Honest question - I don't remember if you actually can use the `PuppetServerConfigService` for this)

KevinCorcoran replied:
Yes, it should use the `PuppetServerConfigService` and not the `ConfigService` directly.

nwolfe commented on test/unit/puppetlabs/services/master/master_service_test.clj:
```diff
@@ -70,3 +69,52 @@
+           ; any SSL files
+           ca/retrieve-ca-cert! (fn [_ _])
+           ca/retrieve-ca-crl! (fn [_ _])
+           ca/initialize-master-ssl! (fn [_ _ _])]
```
Could you use the disabled CA service instead of the actual CA service + with-redefs?

fpringvaldsen commented:
@nwolfe @KevinCorcoran I've addressed your feedback.

One further question: Do we want this to go into master for the 2.0 release? Or do we want this to go into 1.x/2.x, and thus target it at stable?

nwolfe commented on src/clj/puppetlabs/services/master/master_service.clj:
```diff
@@ -4,25 +4,30 @@
+         settings          (ca/config->master-settings config)
+         jruby-service     (tk-services/get-service this :JRubyPuppetService)
+         upgrade-error     (core/construct-404-error-message jruby-service)
+         product-name      (get-in-config [:product :name])
```
Do we need the `get-in-config`? We're already pulling in the whole `config` and using the standard `(get-in config [:k1 :k2])` in this let block

nwolfe commented on test/unit/puppetlabs/services/master/master_service_test.clj:
```diff
@@ -12,10 +11,12 @@
+    [puppetlabs.services.ca.certificate-authority-disabled-service :refer [certificate-authority-disabled-service]]
+    [puppetlabs.dujour.version-check :as version-check]
+    [me.raynes.fs :as fs]
+    [puppetlabs.puppetserver.certificate-authority :as ca]))
```
The `ca` namespace isn't used anymore now that we've got the disabled CA service, right?

fpringvaldsen replied:
It's used in the other test, which tests that the CA service correctly creates the required files.

nwolfe replied:
Didn't you add it but are not using it anymore? The other tests use the `certificate-authority-service`, not the `ca` namespace.

fpringvaldsen replied:
Oh, you're right, I misunderstood, sorry about that. I'll remove it

fpringvaldsen commented:
@nwolfe I've made the changes based on your feedback.

nwolfe commented:
:+1: nice!
